### PR TITLE
Further follow up from #1359

### PR DESCRIPTION
**Pull request by xeno-by**
Builds upon #1367 to address some of the remaining action items in #1362.

### CODE REVIEW

File: scalameta/interactive/src/main/scala/scala/meta/interactive/InteractiveSemanticdb.scala [hunk 1]
```diff
@@ -46,11 +52,12 @@ object InteractiveSemanticdb {
     * @param filename the name of the source file.
     * @param timeout max number of milliseconds to allow the presentation compiler
     *                to typecheck this file.
+    * @param options -P:semanticdb:* options to influence how the document is built.
```
Comment (olafurpg) [inline scalameta/interactive/src/main/scala/scala/meta/interactive/InteractiveSemanticdb.scala]: Mention explicitly if the `-P:semanticdb:` prefix should be included

File: semanticdb/metacp/src/main/scala/scala/meta/internal/scalacp/Scalacp.scala [hunk 2]
```diff
@@ -0,0 +1,532 @@
...
+          case _ => None
+        }
+      }
+      val stoplevels = toplevels.flatMap(sinfo)
+      val sothers = toplevels.flatMap(spackages).distinct ++ others.flatMap(sinfo)
+      ToplevelInfos(classfile, stoplevels, sothers)
+    }
+  }
+
+  private def spackages(sym: SymbolInfoSymbol): List[s.SymbolInformation] = {
```
Comment (olafurpg) [inline semanticdb/metacp/src/main/scala/scala/meta/internal/scalacp/Scalacp.scala]: Nitpick. I find it a bit weird to make as much as possible private in the internal package. These methods are not protected by any bincompat guarantees.

File: semanticdb/metacp/src/main/scala/scala/meta/internal/metacp/package.scala [hunk 3]
```diff
@@ -0,0 +1,17 @@
+package scala.meta.internal
+
+import java.nio.file._
+import scala.tools.asm._
+import scala.tools.asm.ClassReader._
+import scala.tools.asm.tree._
+
+package object metacp {
+  implicit class XtensionAsmPathOps(path: Path) {
+    def toClassNode: ClassNode = {
```
Comment (olafurpg) [inline semanticdb/metacp/src/main/scala/scala/meta/internal/metacp/package.scala]: Nitpick. I personally think `toX` extension methods should be pure data transformations and not do I/O. I think java.nio and scala.concurrent get this right by making unsafe/blocking operations static methods on objects like `Await.result` and `Files.readAllBytes`.

File: semanticdb/metacp/src/main/scala/scala/meta/internal/scalacp/Scalacp.scala [hunk 4]
```diff
@@ -0,0 +1,532 @@
...
+      val (owner, name) = {
+        transitivePackagePath.split("\\.").toList match {
+          case List(name) if name.nonEmpty => ("", name)
+          case ownerSteps :+ name if name.nonEmpty => (ownerSteps.mkString(".") + ".", name)
+          case _ => sys.error(s"unsupported top-level symbol: $sym")
+        }
+      }
+      s.SymbolInformation(
+        symbol = transitivePackagePath + ".",
+        language = Some(s.Language("Scala")),
```
Comment (olafurpg) [inline semanticdb/metacp/src/main/scala/scala/meta/internal/scalacp/Scalacp.scala]: I think we should consider making `language`  an enum and add a `version: String` field. I find it awkward to hardcode `"Scala"` and `"Java"` all over the place and it opens up a can of worms related to equality (case-sensitivity etc.)

File: semanticdb/metap/src/main/scala/scala/meta/internal/metap/Settings.scala [hunk 5]
```diff
@@ -0,0 +1,31 @@
+package scala.meta.internal.metap
+
+final case class Settings(
+    paths: List[String] = Nil,
+    pretty: Boolean = true,
```
Comment (olafurpg) [inline semanticdb/metap/src/main/scala/scala/meta/internal/metap/Settings.scala]: > every boolean option is an enum screaming to get out
> -- Ólafur Páll Geirsson

```scala
sealed trait OutputFormat
object OutputFormat {
  case object Pretty extends OutputFormat
  case object Proto extends OutputFormat
}
```

Comment (olafurpg) [reply semanticdb/metap/src/main/scala/scala/meta/internal/metap/Settings.scala]: Also helps follow the "Make invalid states unrepresentable" principle

File: semanticdb/metacp/src/main/scala/scala/meta/internal/javacp/Javacp.scala [hunk 6]
```diff
@@ -1,86 +1,64 @@
...
+      access: Int,
+      scope: Scope): Seq[s.SymbolInformation] = {
 
     val buf = ArrayBuffer.empty[s.SymbolInformation]
     val decls = ListBuffer.empty[String]
 
     def addInfo(
         symbol: String,
         kind: s.SymbolInformation.Kind,
-        access: Int,
```
Comment (olafurpg) [inline semanticdb/metacp/src/main/scala/scala/meta/internal/javacp/Javacp.scala]: The ordering here was already following the ordering in semanticdb3.proto, `access` is used to set properties which come after kind 😉 The name "access" comes from asm, not semanticdb "accessibility" (which should maybe be called "visibility"?).

File: semanticdb/metacp/src/main/scala/scala/meta/internal/javacp/Javacp.scala [hunk 7]
```diff
@@ -1,86 +1,64 @@
...
 
-  def sinfosClassNode(
+  private def sinfos(
+      toplevel: ToplevelClassfile,
       node: ClassNode,
-      outerClassAccess: Int,
-      scope: Scope,
-      root: Path,
-      isVisited: mutable.Set[Path]): Seq[s.SymbolInformation] = {
+      access: Int,
```
Comment (olafurpg) [inline semanticdb/metacp/src/main/scala/scala/meta/internal/javacp/Javacp.scala]: I intentionally renamed this from access to outerClassAccess in hope to make it clearer. No objections from me if you prefer it this way.

File: semanticdb/metacp/src/main/scala/scala/meta/internal/javacp/Javacp.scala [hunk 8]
```diff
@@ -92,9 +70,9 @@ object Javacp {
           addInfo(
             pkgSymbol,
             k.PACKAGE,
-            o.ACC_PUBLIC,
             pkgName,
             None,
+            o.ACC_PUBLIC,
```
Comment (olafurpg) [inline semanticdb/metacp/src/main/scala/scala/meta/internal/javacp/Javacp.scala]: NB. I'm not sure if "Some(PUBLIC)" is the right accessibility for packages. Maybe it should be None instead. Doesn't matter for me.

File: semanticdb/metacp/src/main/scala/scala/meta/internal/scalacp/Scalacp.scala (hunk 4 above)
Comment (xeno-by) [reply semanticdb/metacp/src/main/scala/scala/meta/internal/scalacp/Scalacp.scala]: Yep, I was thinking about that yesterday. I've opened a ticket: https://github.com/scalameta/scalameta/issues/1374.

File: semanticdb/metacp/src/main/scala/scala/meta/internal/javacp/Javacp.scala (hunk 8 above)
Comment (xeno-by) [reply semanticdb/metacp/src/main/scala/scala/meta/internal/javacp/Javacp.scala]: We already have a related ticket: https://github.com/scalameta/scalameta/issues/1325. Let's follow up there.

File: semanticdb/metacp/src/main/scala/scala/meta/internal/javacp/Javacp.scala (hunk 7 above)
Comment (xeno-by) [reply semanticdb/metacp/src/main/scala/scala/meta/internal/javacp/Javacp.scala]: I think that `outerClassAccess` is somewhat misleading. It's not `access` of an outer class, but `access` of an `InnerClass` entry in outer class. I don't have better ideas though, so let's use `access` for the time being if you don't mind.

Comment (olafurpg) [reply semanticdb/metacp/src/main/scala/scala/meta/internal/javacp/Javacp.scala]: I agree.

File: semanticdb/metacp/src/main/scala/scala/meta/internal/javacp/Javacp.scala (hunk 6 above)
Comment (xeno-by) [reply semanticdb/metacp/src/main/scala/scala/meta/internal/javacp/Javacp.scala]: Uh oh. My bad.

File: semanticdb/metap/src/main/scala/scala/meta/internal/metap/Settings.scala (hunk 5 above)
Comment (xeno-by) [reply semanticdb/metap/src/main/scala/scala/meta/internal/metap/Settings.scala]: 👍

File: semanticdb/metacp/src/main/scala/scala/meta/internal/scalacp/Scalacp.scala (hunk 2 above)
Comment (xeno-by) [reply semanticdb/metacp/src/main/scala/scala/meta/internal/scalacp/Scalacp.scala]: This emphasizes that the methods aren't part of the interface between metacp, javacp and scalacp. However if you'd like, I can make them public again.

Comment (olafurpg) [reply semanticdb/metacp/src/main/scala/scala/meta/internal/scalacp/Scalacp.scala]: I hit on myself the other day that I wanted to access a private method in Metap from external repo. In this particular case it's probably better to come up with a better public API for metap than un-private methods. Nothing to act upon now.

**Pull request merged**